### PR TITLE
Use a single lock for all core classes.

### PR DESCRIPTION
**Pull request by frank-weinberg**
In order to avoid potential for deadlocks, cover all the core logic in
Default*Model with a single lock. This could make reaction times a bit
slower, if multiple clients send state updates at the same time but as
long as all updates are handled quickly, this should be negligible. (We
should, however, not start to compute complex stats under the lock.)

Also make DefaultScoreBoardModel deliver events synchronously. 

The only remaining calls from code under the lock to code (potentially)
under different locks are to the baseLock in Ruleset.java, which in turn
does not reach other locks, and through the event interface. Since all
event listeners not covered by the coreLock are async and thus the
actual handling of the event will not be done while holding the
coreLock, we can conclude that this change and #149 together do not
introduce any potential deadlocks.

### CODE REVIEW

File: src/com/carolinarollergirls/scoreboard/defaults/DefaultClockModel.java [hunk 1]
```diff
@@ -242,45 +263,53 @@ public void setMinimumTime(long ms) {
 		}
 	}
 	public void changeMinimumTime(long change) {
-		synchronized (timeLock) {
+		synchronized (scoreBoardModel) {
```
Comment (brian-brazil) [inline src/com/carolinarollergirls/scoreboard/defaults/DefaultClockModel.java]: coreLock

File: src/com/carolinarollergirls/scoreboard/defaults/DefaultClockModel.java [hunk 2]
```diff
@@ -242,45 +263,53 @@ public void setMinimumTime(long ms) {
...
+		synchronized (coreLock) {
 			setMaximumTime(maximumTime + change);
 		}
 	}
 	public boolean isTimeAtStart(long t) {
-		if (isCountDirectionDown())
-			return t == getMaximumTime();
-		else
-			return t == getMinimumTime();
+		synchronized (scoreBoardModel) {
```
Comment (brian-brazil) [inline src/com/carolinarollergirls/scoreboard/defaults/DefaultClockModel.java]: coreLock

File: src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java [hunk 3]
```diff
@@ -568,44 +581,24 @@ protected void createTeamModel(String id) {
 		scoreBoardChange(new ScoreBoardEvent(this, EVENT_ADD_TEAM, model, null));
 	}
 
-	// Have all events delivered by the ScoreBoard asynchronously.
```
Comment (brian-brazil) [inline src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:572]: This should be mentioned in the change description.

**Review by brian-brazil — COMMENTED** [review]
Is there any reason not to move ruleSet over to coreLock? It seems a bit odd that it's different.

**Comment by frank-weinberg** [comment]
In my perception Ruleset is not part of the scoreboard core. The core classes (Default*Model) maintain the state of the scoreboard, Ruleset does not contain any state.

Everything else is addressed.

**Comment by brian-brazil** [comment]
Well I'll be moving RuleSet into the core when I'm cleaning that up in a bit anyway.

**Pull request merged**